src/rust/src/tracing_ffi.rs

```diff
@@ -62,72 +63,96 @@ pub extern "C" fn tracing_init(
 
-        #[cfg(target_os = "windows")]
-        let log_path = OsString::from_wide(log_path);
+    #[cfg(target_os = "windows")]
+    let log_path = log_path.map(OsString::from_wide);
 
-        tracing_init_file(Path::new(&log_path), initial_filter, log_timestamps)
-    }
+    tracing_init_inner(
+        log_path.as_ref().map(Path::new),
+        initial_filter,
+        log_timestamps,
+    )
 }
 
-fn tracing_init_stdout(initial_filter: &str, log_timestamps: bool) -> *mut TracingHandle {
-    let builder = tracing_subscriber::fmt()
-        .with_ansi(true)
-        .with_env_filter(initial_filter);
-
-    let reload_handle = if log_timestamps {
-        let builder = builder.with_filter_reloading();
-        let reload_handle = builder.reload_handle();
-        builder.init();
-        Box::new(reload_handle) as Box<dyn ReloadHandle>
-    } else {
-        let builder = builder.without_time().with_filter_reloading();
-        let reload_handle = builder.reload_handle();
-        builder.init();
-        Box::new(reload_handle) as Box<dyn ReloadHandle>
-    };
-
-    Box::into_raw(Box::new(TracingHandle {
-        _file_guard: None,
-        reload_handle,
-    }))
-}
-
-fn tracing_init_file(
-    log_path: &Path,
+fn tracing_init_inner(
+    log_path: Option<&Path>,
     initial_filter: &str,
     log_timestamps: bool,
 ) -> *mut TracingHandle {
-    let file_appender =
-        tracing_appender::rolling::never(log_path.parent().unwrap(), log_path.file_name().unwrap());
-    let (non_blocking, file_guard) = tracing_appender::non_blocking(file_appender);
-
-    let builder = tracing_subscriber::fmt()
-        .with_ansi(false)
-        .with_env_filter(initial_filter)
-        .with_writer(non_blocking);
-
-    let reload_handle = if log_timestamps {
-        let builder = builder.with_filter_reloading();
-        let reload_handle = builder.reload_handle();
-        builder.init();
-        Box::new(reload_handle) as Box<dyn ReloadHandle>
+    let (file_logger, file_guard) = if let Some(log_path) = log_path {
+        let file_appender = tracing_appender::rolling::never(
+            log_path.parent().unwrap(),
+            log_path.file_name().unwrap(),
+        );
+        let (non_blocking, file_guard) = tracing_appender::non_blocking(file_appender);
+
+        (
+            Some(
+                tracing_subscriber::fmt::layer()
+                    .with_ansi(false)
+                    .with_writer(non_blocking),
+            ),
+            Some(file_guard),
+        )
     } else {
-        let builder = builder.without_time().with_filter_reloading();
-        let reload_handle = builder.reload_handle();
-        builder.init();
-        Box::new(reload_handle) as Box<dyn ReloadHandle>
+        (None, None)
+    };
+    let stdout_logger = tracing_subscriber::fmt::layer().with_ansi(true);
+    let filter = EnvFilter::from(initial_filter);
+
+    let reload_handle = match (file_logger, log_timestamps) {
```

Comment:
This `match` is required due to the need to separately build and initialize every variant, due to their differing types. Once https://github.com/tokio-rs/tracing/issues/894 merges, we can collapse almost all of this into a single tracing subscriber that optionally enables most functionality. (The one part that might not collapse is `log_timestamps`, as that is modifying a layer, not adding a new one, but we could maybe model that as two alternative layers in the type system.)

Reply:
Opened #4668.